workflow/controller/operator.go

```diff
@@ -2587,6 +2601,24 @@ func (woc *wfOperationCtx) addChildNode(parent string, child string) {
 	woc.updated = true
 }
 
+// getFailHosts returns slice of all child nodes with fail or error status
+func (woc *wfOperationCtx) getFailHosts(parent string) []string {
```

Comment:
(a) this is generic and so I don't think it should a utility func in the already bloated and hard to test woc.

(b) so I think this could be refactored as follows:

```golang
// Filter returns the children of the parent. 
func (s Nodes) Children(parentNodeId string) Nodes {...}
```


```golang
// Filter returns the subset of the nodes that match the predicate, e.g. only failed nodes
func (s Nodes) Filter(predicate func (NodeStatus) bool) Nodes {...}
```

```golang
// Map maps the nodes to new values, e.g. `x.Hostname`
func (s Nodes) Map(func (x NodeStatus) interface{}) map[string]interface{} {...}
```

```golang
var hostNames []string
for _, hostName := range woc.wf.Status.Nodes.
  Children(parent).
  Filter(func(x NodeStatus) {return x.Phase == NodeFailed || x.Phase == NodeError}).
  Map(func (x NodeStatus) {return x.Hostname)) {
    hostNames = append(hostNames, hostName.(string)
}
```